<h1 align='center'> INF-280 Estadística Computacional 2023-1 </h1>

<H3 align='center'> Tema 7 y 8 - Intervalos de Confianza y Test de Hipótesis  </H3>

### **Datos del grupo**



 Nombre del grupo: Los dESTACAdos

*   Maximiliano Bardi - 202173510-0 - [video](https://www.youtube.com/watch?v=V-8-WuZQ8sw&ab_channel=MaxBardi)
*   Javiera Bobadilla - 202173584-4 - [video](https://drive.google.com/file/d/1Xfc8cDMQ7YkFtS-DtArOFDsM6-V0flSr/view?usp=sharing)
*   Javiera Osorio Mardones - 202173656-5 - [video](https://drive.google.com/file/d/1OGXOo2J0YnNqnJ39L9CRe9ceTBJGMNmR/view?usp=sharing)

*Recuerden que si su grupo se llama "Llamado de emergencia", entonces el nombre del archivo debe ser T78_Llamado_de_emergencia.ipynb*

### **Formalidades**  

* Es posible utilizar apuntes, libros, papers, ejemplos y cualquier otro material que desee de internet, pero deben ser ustedes quienes desarrollen los ejercicios y comenten los resultados en el Notebook que entregará.
* Se permite el uso de las librerías: pandas, numpy, matplotlib y seaborn. Si desea utilizar otra librería debe indicar cómo instalarla.
* Se debe trabajar en grupos de **dos o tres personas**.
* Además del Notebook, deberán entregar un video explicando lo que hizo, paso a paso, como en los ejemplos entregados en AULA. Este vídeo es **individual**, es decir, cada integrante debe hacerlo por separado (debiendo utilizar el mismo Jupyter Notebook). El video puede ser subido a una plataforma externa como Youtube, Google Drive, ó Dropbox. El **link a compartir** (*visible para cualquier usuario con el link*) debe ir dentro del Notebook.
* La nota de esta actividad es grupal.

### **Descuentos**

* Se descontarán 5 puntos si el nombre del archivo subido no sigue el formato indicado.
* Se descontarán 5 puntos por cada hora de atraso en la entrega.
* Se descontarán 3 puntos por cada gráfico que no incluya los nombres de los ejes.

### **Entrega**

* **Formato**: Se debe realizar una **única** entrega por grupo (un representate del grupo) a través de  __[**Aula**](https://aula.usm.cl/)__.
* **Archivo**:

    *   El nombre del archivo a subir debe ser **T78_Nombre_Del_Equipo.ipynb**, por ejemplo, si su equipo se llama "Llamado de emergencia", entonces el nombre del archivo debe ser **T78_Llamado_de_emergencia.ipynb**
    *   El Jupyter Notebook debe incluir **todos links** a los vídeos explicativos, uno por cada integrante.
    *   Cada video debe tener una **duración máxima de 15 minutos**.
* **Fecha límite**: Domingo 2 de Julio hasta las 23:59.

<img src='https://capitalhumano.emol.com/wp-content/uploads/2019/01/salarychief.jpg' width='500' style='display:block; margin:auto;'/>

## Contexto

En el campo de la ingeniería de software, es importante analizar si existen diferencias salariales significativas basadas en factores como el género, la experiencia laboral y la educación. Esto nos ayudará a comprender posibles brechas salariales y desigualdades en esta industria altamente competitiva. Al explorar estas variables, podremos evaluar la igualdad salarial y las oportunidades profesionales para hombres y mujeres en el campo de la ingeniería de software.

## Preguntas

**Tema 7: Intervalos de Confianza**

**Importante** para las siguientes preguntas:
* Indique explicitamente **los supuestos** que consideraron.Esto implica especificar si se asumió que la varianza poblacional es conocida o no. Además, es fundamental proporcionar una justificación clara y fundamentada para respaldar la elección de ese supuesto en particular.
* Si realizan un intervalo de confianza (IC) para la diferencia de medias, es importante indicar si se consideró la suposición de homocedasticidad o heterocedasticidad. Esto implica especificar si se asumió que las varianzas de las dos poblaciones son iguales (homocedasticidad) o si se permitió que fueran diferentes (heterocedasticidad).


1. **[20 pts]** Construya un Intervalo de Confianza para el salario ("Salary") promedio de las personas cuyo título profesional es "Software Engineer". ¿Es correcto afirmar que el **salario promedio verdadero** de los "Software Engineer" es alrededor de \$113.000 ?. En caso de que sea correcto, ¿Que ocurre si se tuviese una muestra con la misma media y desviación estándar, pero con 100 veces más datos?, es decir, ¿Que ocurre con tamaño del intervalo de confianza?, ¿Se puede seguir afirmando que el salario promedio verdadero es alrededor de \\$113.000 ?

2. **[15 pts]** Compare el salario promedio de los "Software Engineer" que tienen **3 o más** años de experiencia con aquellos que tienen **menos de 3 años** de experiencia. ¿Es correcto afirmar que la **diferencia media verdadera** entre sus sueldos es alrededor de \$65.000?

3. **[15 pts]** Se piensa que existe una brecha salarial de género en las personas que son "Senior Software Engineer". ¿Es correcto afirmar que la **diferencia media verdadera** entre los sueldos de los hombres v/s mujeres es alrededor de \$20.000?

---
**Tema 8: Test de Hipótesis**

**Importante** para las siguientes preguntas:
* Indique explicitamente **los supuestos** que consideraron.
* Indique explicitamente **$H_0$** y **$H_1$**.
* Indique explicitamente el valor del **Estadístico de Contraste**.
* Indique explicitamente la **Región Crítica**.
* Indique explicitamente si **se rechaza $H_0$** o **no se rechaza $H_0$**.

4. **[20 pts]** Se piensa que existe una brecha salarial de género en las personas que son "Senior Software Engineer". Diseñe un contraste que permita evaluar la hipótesis de que "el salario promedio de las mujeres que son 'Senior Software Engineer' **es menor** que el de los hombres con el mismo título". Diseñe su contraste de manera que la probabilidad de aceptar incorrectamente esta conjetura es menor al 5%. Además, **determine el p-valor** e indique que ocurre para valores de $\alpha$ mayores y menores al p-valor.

5. **[15 pts]** Diseñe un contraste que permita evaluar la hipótesis de que "*la proporción de  mujeres que son 'Software Engineer' **es menor** que la proporción de hombres que son 'Software Engineer'*". Diseñe su contraste de manera que la probabilidad de aceptar incorrectamente esta conjetura es menor al 5%. ¿Qué se puede concluir?

6. **[15 pts]** Se piensa que existe una brecha de género en las personas con educación superior ('Bachelor's Degree', 'Master's Degree', 'PhD'). Ejecute un contraste donde $H_0$ corresponda a que "*la proporción de mujeres con educación superior es **mayor o igua**l a 0.5*". Considere $\alpha = 0.05$. Además, **determine el p-valor** e interprete este valor, ¿Qué se puede concluir?

## Librerías

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns; sns.set()
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")

## Lectura de Datos

In [ ]:
df_salary = pd.read_csv('https://drive.google.com/uc?id=1JJBea7ezpNqzOaDYv4OTsjL8VD3J9bmn')

## Funciones útiles

### Funciones para Intervalos de Confianza

A continuación se les entregan una serie de funciones que les podrían servir para realizar el taller. Se recomienda que las revisen por si es que existe algún error.

In [ ]:
from scipy.stats import norm, t, chisquare
from scipy.stats import f as fisher


def IC_media(n, media_muestral, std_muestral, alpha=0.05):
    return t.interval(1-alpha, df=n-1, loc=media_muestral, scale=std_muestral/np.sqrt(n))

def IC_prop(n1,frac1,alpha=0.05):
  std_err = np.sqrt((frac1*(1.0-frac1)/n1))
  z_alpha = norm.ppf(1-alpha/2.0)
  IC_inf = frac1 - z_alpha * std_err
  IC_sup = frac1 + z_alpha * std_err
  return (IC_inf,IC_sup)

#para media1-media2
def IC_dos_medias_hetero(muestra1, muestra2, alpha=0.05):
  n1 = len(muestra1); n2 = len(muestra2);
  media1 = np.mean(muestra1); media2 = np.mean(muestra2);
  std1 = np.std(muestra1,ddof=1); std2 = np.std(muestra2,ddof=1);
  gl_num = (std1**2/float(n1) + std2**2/float(n2))**2
  gl_den = ((std1**2/float(n1))**2)/(n1-1.0) + ((std2**2/float(n2))**2)/(n2-1.0)
  gl = int(np.floor(gl_num/gl_den))
  t_alpha = t.ppf(1-alpha/2.0,df=gl)
  std_error = np.sqrt(std1**2/n1 + std2**2/n2)
  #print(">>",gl,std_error,t_alpha)
  IC_inf = (media1-media2) - t_alpha * std_error
  IC_sup = (media1-media2) + t_alpha * std_error
  return IC_inf,IC_sup

#para prop1-prop2
def IC_dos_props(n1,n2,frac1,frac2, alpha=0.05):
  std_err = np.sqrt((frac1*(1.0-frac1)/n1) + (frac2*(1.0-frac2)/n2))
  z_alpha = norm.ppf(1-alpha/2.0)
  IC_inf = (frac1-frac2) - z_alpha * std_err
  IC_sup = (frac1-frac2) + z_alpha * std_err
  return IC_inf,IC_sup

### Funciones para Test de Hipótesis

In [ ]:
from scipy.stats import norm, t
from scipy.stats import f as fisher


#H0: mu1 = mu0 (mu1 desconocida, m0 valor fijo, sigma desconocida)
def contraste_1media(n1,media1,std1,mu0,error_tipo1,alternativa='1>0'):
  region_critica = []
  gl = n1-1
  std_error = np.sqrt(std1**2/n1)
  T0 = (media1-mu0)/std_error#estadistico bajo H0
  if (alternativa=='1>0'):
    region_critica.append(pd.Interval(left=t.ppf(1-error_tipo1,df=gl), right=np.inf,closed='neither'))
    p_value = 1-t.cdf(T0,df=gl)
  elif (alternativa=='1<0'):
    region_critica.append(pd.Interval(left=-np.inf, right=t.ppf(error_tipo1,df=gl),closed='neither'))
    p_value = t.cdf(T0,df=gl)
  else:#dos colas
    region_critica.append(pd.Interval(left=-np.inf,right=t.ppf(error_tipo1/2,df=gl),closed='neither'))
    region_critica.append(pd.Interval(left=t.ppf(1-error_tipo1/2,df=gl),right=np.inf,closed='neither'))
    p_value = 2*min(1-t.cdf(T0,df=gl),t.cdf(T0,df=gl))
  return T0, region_critica, p_value


#H0: mu_x = mu_y (mu_x,mu_y ambas no conocidas, sigma diferente y desconocido)
def contraste_2medias(n_x,n_y,media_x,media_y,std_x,std_y,error_tipo1,alternativa='x>y'):
  region_critica = []
  gl_num = (std_x**2/float(n_x) + std_y**2/float(n_y))**2
  gl_den = ((std_x**2/float(n_x))**2)/(n_x-1.0) + ((std_y**2/float(n_y))**2)/(n_y-1.0)
  gl = int(np.floor(gl_num/gl_den))
  std_error = np.sqrt(std_x**2/n_x + std_y**2/n_y)
  T0 = (media_x-media_y)/std_error#estadistico bajo H0
  if (alternativa=='x>y'):
    region_critica.append(pd.Interval(left=t.ppf(1-error_tipo1,df=gl), right=np.inf,closed='neither'))
    p_value = 1-t.cdf(T0,df=gl)
  elif (alternativa=='x<y'):
    region_critica.append(pd.Interval(left=-np.inf, right=t.ppf(error_tipo1,df=gl),closed='neither'))
    p_value = t.cdf(T0,df=gl)
  else:#dos colas
    region_critica.append(pd.Interval(left=-np.inf,right=t.ppf(error_tipo1/2,df=gl),closed='neither'))
    region_critica.append(pd.Interval(left=t.ppf(1-error_tipo1/2,df=gl),right=np.inf,closed='neither'))
    p_value = 2*min(1-t.cdf(T0,df=gl),t.cdf(T0,df=gl))
  print(gl)
  return T0, region_critica, p_value

#H0: sigma1 = sigma2 (sigma1,sigma2 ambas no conocidas)
def contraste_dos_varianzas(n_x,n_y,std_x,std_y,error_tipo1,alternativa='x>y'):
  region_critica = []
  F0 = std_x**2/std_y**2
  if (alternativa=='x>y'):
    region_critica.append(pd.Interval(left=fisher.ppf(1-error_tipo1,dfn=n_x-1, dfd=n_y-1), right=np.inf,closed='neither'))
    p_value = 1-fisher.cdf(F0,dfn=n_x-1, dfd=n_y-1)
  elif (alternativa=='x<y'):
    region_critica.append(pd.Interval(left=-np.inf, right=fisher.ppf(error_tipo1,dfn=n_x-1, dfd=n_y-1),closed='neither'))
    p_value = fisher.cdf(F0,dfn=n_x-1, dfd=n_y-1)
  else:#dos colas
    region_critica.append(pd.Interval(left=-np.inf,right=fisher.ppf(error_tipo1/2,dfn=n_x-1, dfd=n_y-1),closed='neither'))
    region_critica.append(pd.Interval(left=fisher.ppf(1-error_tipo1/2,dfn=n_x-1, dfd=n_y-1),right=np.inf,closed='neither'))
    p_value = 2*min(1-fisher.cdf(F0,dfn=n_x-1, dfd=n_y-1),fisher.cdf(F0,dfn=n_x-1, dfd=n_y-1))
  return F0, region_critica, p_value

#H0: p1 = p0 (p1 desconocida, p0 valor fijo)
def contraste_1prop(n1,prop1,p0,error_tipo1,alternativa='1>0'):
  region_critica = []
  std_error = np.sqrt(p0*(1.0-p0))*np.sqrt(1.0/n1)
  Z0 = (prop1-p0)/std_error
  if (alternativa=='1>0'):
    region_critica.append(pd.Interval(left=norm.ppf(1-error_tipo1), right=np.inf,closed='neither'))
    p_value = 1-norm.cdf(Z0)
  elif (alternativa=='1<0'):
    region_critica.append(pd.Interval(left=-np.inf, right=norm.ppf(error_tipo1),closed='neither'))
    p_value = norm.cdf(Z0)
  else:#dos colas
    region_critica.append(pd.Interval(left=-np.inf,right=norm.ppf(error_tipo1/2),closed='neither'))
    region_critica.append(pd.Interval(left=norm.ppf(1-error_tipo1/2),right=np.inf,closed='neither'))
    print(">>",Z0,1-norm.cdf(Z0),norm.cdf(Z0))
    p_value = 2*min(1-norm.cdf(Z0),norm.cdf(Z0))
  return Z0, region_critica, p_value

#H0: p1 = p2 (p1,p2 no conocidas, ni funcion una de la otra)
def contraste_2props(n_x,n_y,prop_x,prop_y,error_tipo1,alternativa='x>y'):
  region_critica = []
  prop_comb = (n_x*prop_x + n_y*prop_y)/(n_x+n_y)
  std_error = np.sqrt(prop_comb*(1.0-prop_comb))*np.sqrt(1.0/n_x + 1.0/n_y)
  Z0 = (prop_x-prop_y)/std_error

  if (alternativa=='x>y'):
    region_critica.append(pd.Interval(left=norm.ppf(1-error_tipo1), right=np.inf,closed='neither'))
    p_value = 1-norm.cdf(Z0)
  elif (alternativa=='x<y'):
    region_critica.append(pd.Interval(left=-np.inf, right=norm.ppf(error_tipo1),closed='neither'))
    p_value = norm.cdf(Z0)
  else:#dos colas
    region_critica.append(pd.Interval(left=-np.inf,right=norm.ppf(error_tipo1/2),closed='neither'))
    region_critica.append(pd.Interval(left=norm.ppf(1-error_tipo1/2),right=np.inf,closed='neither'))
    print(">>",Z0,1-norm.cdf(Z0),norm.cdf(Z0))
    p_value = 2*min(1-norm.cdf(Z0),norm.cdf(Z0))
  return Z0, region_critica, p_value

## Ejemplo Test de Hipótesis:


En un test de hipótesis, la hipótesis nula ($H_0$) corresponde al *status quo*, es decir, a lo que se "cree que es lo normal". En cambio, la hipótesis alternativa es lo que se "sospecha" que ocurre.

Ejemplo:

Diseñe un contraste que permita evaluar la hipótesis de que "la media de la población X es distinta a la media de la población Y". Diseñe un contraste de manera que la probabilidad de aceptar incorrectamente esta conjetura sea menor al 5%.

En este caso, se tiene que:
* $H_0$: $μ_X = \mu_Y$ (Corresponde a lo que se cree "normal")
* $H_1$: $μ_X \neq \mu_Y$

Además, se plantean de esta manera las hipótesis ya que "aceptar incorrectamente la conjetura" corresponde a un error tipo 1 (valor de $α$) y se sabe que siempre se debe acotar este tipo de error, pues es más grave que un error tipo 2.

## Desarrollo

### 1)

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns; sns.set()
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")
from scipy.stats import norm, t, chisquare
from scipy.stats import f as fisher



df_salary = pd.read_csv('https://drive.google.com/uc?id=1JJBea7ezpNqzOaDYv4OTsjL8VD3J9bmn')

def IC_media(n, media_muestral, std_muestral, alpha=0.05):
    return t.interval(1-alpha, df=n-1, loc=media_muestral, scale=std_muestral/np.sqrt(n))

soft_E= df_salary[df_salary['Job Title'] == 'Software Engineer']
salarios = soft_E['Salary']

media_muestral = np.mean(salarios)
std_muestral = np.std(salarios, ddof=1)

n = len(salarios)

print("Intervalo de confianza: ",IC_media(n, media_muestral, std_muestral, alpha=0.05  ))
print("Salario promedio verdadero: ", media_muestral)

Intervalo de confianza:  (109097.28115071809, 117389.20533576839)
Salario promedio verdadero:  113243.24324324324


Si es que toda la población de los que trabajan como Software Engineere se encontrase en ese excel, entonces su promedio verdadero correspondería a la media muestral de los datos, la cual da un valor alrededor de 113.000 (truncado) por lo que es correcto afirmar en este caso hipotético que 113.000 corresponde a la media verdadera.

Si es que se aumentan los datos, el intervalo de confianza se haría más pequeño, ya que a mayor presición menor es la cantidad de los números "entre medio" del dato ideal.

Si es que se aumentan las datos existe la posibilidad que la media verdadera cambie (si es que dentro de estos datos hay una tendencia a un valor mayor o menor que la media estipulada) por lo que no es muy certero seguir afirmando que el promedio verdadero sea 113000.


### 2)

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns; sns.set()
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")
from scipy.stats import norm, t, chisquare
from scipy.stats import f as fisher



df_salary = pd.read_csv('https://drive.google.com/uc?id=1JJBea7ezpNqzOaDYv4OTsjL8VD3J9bmn')

soft_E= df_salary[df_salary['Job Title'] == 'Software Engineer']

softE_mayor = soft_E[soft_E["Years of Experience"] >= 3]
softE_menor = soft_E[soft_E["Years of Experience"] < 3]

salario_M = softE_mayor['Salary']
salario_m = softE_menor['Salary']


media_muestralM = np.mean(salario_M)
media_muestralm = np.mean(salario_m)

std_muestralM = np.std(salario_M, ddof=1)
std_muestralm = np.std(salario_m, ddof=1)

n_M = len(salario_M)
n_m = len(salario_m)

def IC_dos_medias_hetero(muestra1, muestra2, alpha=0.05):
  n1 = len(muestra1); n2 = len(muestra2);
  media1 = np.mean(muestra1); media2 = np.mean(muestra2);
  std1 = np.std(muestra1,ddof=1); std2 = np.std(muestra2,ddof=1);
  gl_num = (std1**2/float(n1) + std2**2/float(n2))**2
  gl_den = ((std1**2/float(n1))**2)/(n1-1.0) + ((std2**2/float(n2))**2)/(n2-1.0)
  gl = int(np.floor(gl_num/gl_den))
  t_alpha = t.ppf(1-alpha/2.0,df=gl)
  std_error = np.sqrt(std1**2/n1 + std2**2/n2)
  #print(">>",gl,std_error,t_alpha)
  IC_inf = (media1-media2) - t_alpha * std_error
  IC_sup = (media1-media2) + t_alpha * std_error
  return IC_inf,IC_sup




print("Intervalo de confianza entre las dos muestras: ", IC_dos_medias_hetero(salario_M, salario_m,))

Intervalo de confianza entre las dos muestras:  (62261.262758844314, 71108.91030360867)


Es posible decir que la media verdadera entre ambos se encuentra alrededor de 65.000 según los datos entregados.

### 3)

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns; sns.set()
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")
from scipy.stats import norm, t, chisquare
from scipy.stats import f as fisher



df_salary = pd.read_csv('https://drive.google.com/uc?id=1JJBea7ezpNqzOaDYv4OTsjL8VD3J9bmn')

soft_E= df_salary[df_salary['Job Title'] == 'Senior Software Engineer']

softE_h = soft_E[soft_E["Gender"] == "Male"]
softE_m = soft_E[soft_E["Gender"] == "Female"]

salario_h = softE_h['Salary']
salario_m = softE_m['Salary']


def IC_dos_medias_hetero(muestra1, muestra2, alpha=0.05):
  n1 = len(muestra1); n2 = len(muestra2);
  media1 = np.mean(muestra1); media2 = np.mean(muestra2);
  std1 = np.std(muestra1,ddof=1); std2 = np.std(muestra2,ddof=1);
  gl_num = (std1**2/float(n1) + std2**2/float(n2))**2
  gl_den = ((std1**2/float(n1))**2)/(n1-1.0) + ((std2**2/float(n2))**2)/(n2-1.0)
  gl = int(np.floor(gl_num/gl_den))
  t_alpha = t.ppf(1-alpha/2.0,df=gl)
  std_error = np.sqrt(std1**2/n1 + std2**2/n2)
  #print(">>",gl,std_error,t_alpha)
  IC_inf = (media1-media2) - t_alpha * std_error
  IC_sup = (media1-media2) + t_alpha * std_error
  return IC_inf,IC_sup



print("Intervalo de confianza entre las dos muestras: ", IC_dos_medias_hetero(salario_h, salario_m, alpha=0.05))


Intervalo de confianza entre las dos muestras:  (16549.104818884014, 30374.78658383087)


La media verdadera efectivamente se encuentra alrededor de 20.000

### 4)

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns; sns.set()
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")
from scipy.stats import norm, t, chisquare
from scipy.stats import f as fisher

df_salary = pd.read_csv('https://drive.google.com/uc?id=1JJBea7ezpNqzOaDYv4OTsjL8VD3J9bmn')

soft_E= df_salary[df_salary['Job Title'] == 'Senior Software Engineer']

softE_h = soft_E[soft_E["Gender"] == "Male"]
softE_m = soft_E[soft_E["Gender"] == "Female"]

salario_h = softE_h['Salary']
salario_m = softE_m['Salary']


media_muestralh = np.mean(salario_h)
media_muestralm = np.mean(salario_m)

std_muestralh = np.std(salario_h, ddof=1)
std_muestralm = np.std(salario_m, ddof=1)

n_h = len(salario_h)
n_m = len(salario_m)


def contraste_2medias(n_x,n_y,media_x,media_y,std_x,std_y,error_tipo1,alternativa='x>y'):
  region_critica = []
  gl_num = (std_x**2/float(n_x) + std_y**2/float(n_y))**2
  gl_den = ((std_x**2/float(n_x))**2)/(n_x-1.0) + ((std_y**2/float(n_y))**2)/(n_y-1.0)
  gl = int(np.floor(gl_num/gl_den))
  std_error = np.sqrt(std_x**2/n_x + std_y**2/n_y)
  T0 = (media_x-media_y)/std_error#estadistico bajo H0
  if (alternativa=='x>y'):
    region_critica.append(pd.Interval(left=t.ppf(1-error_tipo1,df=gl), right=np.inf,closed='neither'))
    p_value = 1-t.cdf(T0,df=gl)
  elif (alternativa=='x<y'):
    region_critica.append(pd.Interval(left=-np.inf, right=t.ppf(error_tipo1,df=gl),closed='neither'))
    p_value = t.cdf(T0,df=gl)
  else:#dos colas
    region_critica.append(pd.Interval(left=-np.inf,right=t.ppf(error_tipo1/2,df=gl),closed='neither'))
    region_critica.append(pd.Interval(left=t.ppf(1-error_tipo1/2,df=gl),right=np.inf,closed='neither'))
    p_value = 2*min(1-t.cdf(T0,df=gl),t.cdf(T0,df=gl))

  return T0, region_critica, p_value


error_tipo1 = 0.05
t_valor, region_critica, p = contraste_2medias(n_h, n_m, media_muestralh, media_muestralm,
                                                     std_muestralh, std_muestralm, error_tipo1)

print("Valor t: ", t_valor)
print("La región crítica es: ", region_critica)
print("Valor p:", p)

Valor t:  6.697077976210162
La región crítica es:  [Interval(1.6533630134964312, inf, closed='neither')]
Valor p: 1.3111289831613249e-10


Para un alfa mayor no podríamos saber si es que existe una brecha, pero con el alfa menor podemos definir con certeza esa brecha (con el valor p)

### 5)

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns; sns.set()
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")
from scipy.stats import norm, t, chisquare
from scipy.stats import f as fisher

df_salary = pd.read_csv('https://drive.google.com/uc?id=1JJBea7ezpNqzOaDYv4OTsjL8VD3J9bmn')

soft_E= df_salary[df_salary['Job Title'] == 'Software Engineer']

softE_h = soft_E[soft_E["Gender"] == "Male"]
softE_m = soft_E[soft_E["Gender"] == "Female"]

mujeres =df_salary[df_salary["Gender"] == "Female"]
ntM = len(mujeres)

hombres =df_salary[df_salary["Gender"] == "Male"]
ntH = len(hombres)

nt = len(soft_E)
n_h = len(softE_h)
n_m = len(softE_m)

proph = n_h/nt
propm = n_m/nt


def contraste_2props(n_x,n_y,prop_x,prop_y,error_tipo1,alternativa='x>y'):
  region_critica = []
  prop_comb = (n_x*prop_x + n_y*prop_y)/(n_x+n_y)
  std_error = np.sqrt(prop_comb*(1.0-prop_comb))*np.sqrt(1.0/n_x + 1.0/n_y)
  Z0 = (prop_x-prop_y)/std_error

  if (alternativa=='x>y'):
    region_critica.append(pd.Interval(left=norm.ppf(1-error_tipo1), right=np.inf,closed='neither'))
    p_value = 1-norm.cdf(Z0)
  elif (alternativa=='x<y'):
    region_critica.append(pd.Interval(left=-np.inf, right=norm.ppf(error_tipo1),closed='neither'))
    p_value = norm.cdf(Z0)
  else:#dos colas
    region_critica.append(pd.Interval(left=-np.inf,right=norm.ppf(error_tipo1/2),closed='neither'))
    region_critica.append(pd.Interval(left=norm.ppf(1-error_tipo1/2),right=np.inf,closed='neither'))
    print(">>",Z0,1-norm.cdf(Z0),norm.cdf(Z0))
    p_value = 2*min(1-norm.cdf(Z0),norm.cdf(Z0))
  return Z0, region_critica, p_value

error_tipo1 = 0.05



Z0, region_critica, p_value = contraste_2props(n_h, n_m, proph, propm, error_tipo1)


print("Z0: ", Z0)
print("Región crítica: ", region_critica)
print("Valor de p: ", p_value)



Z0:  5.620140999457864
Región crítica:  [Interval(1.6448536269514722, inf, closed='neither')]
Valor de p:  9.540083834025381e-09


En este caso tenemos que la hipotesis nula H0 es

*   H0: "la proporción de mujeres que son 'Software Engineer' es menor que la proporción de hombres que son 'Software Engineer'"
*   H1: la proporción de mujeres que son 'Software Engineer' es mayor o igual que la proporción de hombres que son 'Software Engineer'"   
 Estas hipotesis se plantean de esta manera, ya que se nos pide aceptar esta conjetura de manera incorrecta, la cual corresponde a un error de tipo 1 y se sabe que se debe acotar a este tipo de error porque es un error más grave que el de tipo 2

 Dado el resultado que se obtuvo se tiene que el valor de p es de 9.540083834025381e-09, el cual es un valor bastante cercano a 0, lo cual indica que el error de tipo 1 generado es mayor al de tipo 2 pero no llega a ser máximo, por lo que no es el peor error




### 6)

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns; sns.set()
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")
from scipy.stats import norm, t, chisquare
from scipy.stats import f as fisher

df_salary = pd.read_csv('https://drive.google.com/uc?id=1JJBea7ezpNqzOaDYv4OTsjL8VD3J9bmn')


mujeres_E = df_salary[(df_salary['Gender'] == 'Female') & (df_salary['Education Level'].isin(["Bachelor's Degree", "Master's Degree", "PhD"]))]
mujeres =df_salary[df_salary["Gender"] == "Female"]
nt = len(mujeres)
n_me = len(mujeres_E)

prop_m = n_me/nt

def contraste_1prop(n1,prop1,p0,error_tipo1,alternativa='1>0'):
  region_critica = []
  std_error = np.sqrt(p0*(1.0-p0))*np.sqrt(1.0/n1)
  Z0 = (prop1-p0)/std_error
  if (alternativa=='1>0'):
    region_critica.append(pd.Interval(left=norm.ppf(1-error_tipo1), right=np.inf,closed='neither'))
    p_value = 1-norm.cdf(Z0)
  elif (alternativa=='1<0'):
    region_critica.append(pd.Interval(left=-np.inf, right=norm.ppf(error_tipo1),closed='neither'))
    p_value = norm.cdf(Z0)
  else:#dos colas
    region_critica.append(pd.Interval(left=-np.inf,right=norm.ppf(error_tipo1/2),closed='neither'))
    region_critica.append(pd.Interval(left=norm.ppf(1-error_tipo1/2),right=np.inf,closed='neither'))
    print(">>",Z0,1-norm.cdf(Z0),norm.cdf(Z0))
    p_value = 2*min(1-norm.cdf(Z0),norm.cdf(Z0))
  return Z0, region_critica, p_value

p0 = 0.5
error_tipo1 = 0.05
Z0, region_critica, p_value = contraste_1prop(n_h, prop_m, p0, error_tipo1,)


print("Z0: ", Z0)
print("Región crítica: ", region_critica)
print("Valor de p: ", p_value)




Z0:  9.24715041782901
Región crítica:  [Interval(1.6448536269514722, inf, closed='neither')]
Valor de p:  0.0


Para este caso tenemos que la hipotesis nula (H0) es

*    H0: "la proporción de mujeres con educación superior es mayor o igual a 0.5"
*   H1: "la proporción de mujeres con educación superior es menor a 0.5"   


 Dado el resultado que se obtuvo se tiene que el valor de p es cercano a 0, con un intervalo que va de 1.6 hacia el infinito y un z0=9.24715041782901, esto indica que el error de tipo 1 generado es mayor al de tipo 2


